File: sdk/go/sauronid/budget_tracker.go
package sauronid

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"sync"
	"time"
)

// PendingSpendRecord is one queued spend waiting for the next Flush.
type PendingSpendRecord struct {
	AmountUsd float64 `json:"amount_usd"`
	ActionID  string  `json:"action_id,omitempty"`
	// Timestamp is the unix-epoch milliseconds when Record was called.
	Timestamp int64 `json:"timestamp"`
}

// BudgetState is the snapshot handed to a flush function.
type BudgetState struct {
	PolicyID         string
	TotalUsd         float64
	CallTimestampsMs []int64
	Pending          []PendingSpendRecord
}

// FlushFn is invoked from the background timer with the current state.
// Returning a non-nil error preserves the pending list so the next tick
// retries.
type FlushFn func(ctx context.Context, state BudgetState) error

// BudgetTrackerOptions configures a BudgetTracker.
type BudgetTrackerOptions struct {
	// PolicyID this tracker covers.
	PolicyID string
	// FlushInterval is the auto-flush period. Zero disables the timer.
	// Default 30s.
	FlushInterval time.Duration
	// FlushFn is the hook drained by the timer. Nil = silently drop pending.
	FlushFn FlushFn
}

// BudgetTracker is a thread-safe in-memory spend + rate ledger.
//
// Sprint 3 wires an optional server-side flush so the in-memory total is
// no longer the source of truth (closes redteam A3: local counter
// tampering). See ServerPush for the canonical flush builder.
type BudgetTracker struct {
	policyID      string
	flushInterval time.Duration
	flushFn       FlushFn

	mu               sync.Mutex
	totalUsd         float64
	callTimestampsMs []int64
	pending          []PendingSpendRecord
	stopCh           chan struct{}
	stopped          bool
	wg               sync.WaitGroup
}

// NewBudgetTracker constructs a tracker. If FlushInterval > 0 a
// background goroutine is started immediately.
func NewBudgetTracker(opts BudgetTrackerOptions) *BudgetTracker {
	if opts.FlushInterval == 0 {
		opts.FlushInterval = 30 * time.Second
	}
	b := &BudgetTracker{
		policyID:      opts.PolicyID,
		flushInterval: opts.FlushInterval,
		flushFn:       opts.FlushFn,
		stopCh:        make(chan struct{}),
	}
	if opts.FlushInterval > 0 {
		b.wg.Add(1)
		go b.loop()
	}
	return b
}

// Record adds one tool invocation to the ledger. amountUsd is added to
// the running total (use 0 for no-spend calls) and a unix-millisecond
// timestamp is appended for the rate-window.
func (b *BudgetTracker) Record(amountUsd float64, actionID string) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.totalUsd += amountUsd
	now := time.Now().UnixMilli()
	b.callTimestampsMs = append(b.callTimestampsMs, now)
	// Cap history at last 1024 entries — matches TS bound.
	if len(b.callTimestampsMs) > 1024 {
		b.callTimestampsMs = b.callTimestampsMs[len(b.callTimestampsMs)-1024:]
	}
	b.pending = append(b.pending, PendingSpendRecord{
		AmountUsd: amountUsd,
		ActionID:  actionID,
		Timestamp: now,
	})
}

// Total returns the current spend total in USD.
func (b *BudgetTracker) Total() float64 {
	b.mu.Lock()
	defer b.mu.Unlock()
	return b.totalUsd
}

// PendingCount returns the number of records waiting for the next flush.
func (b *BudgetTracker) PendingCount() int {
	b.mu.Lock()
	defer b.mu.Unlock()
	return len(b.pending)
}

// RecentCalls returns the call timestamps (unix-epoch milliseconds)
// inside the last `window` duration. Older entries are pruned as a
// side effect.
func (b *BudgetTracker) RecentCalls(window time.Duration) []int64 {
	b.mu.Lock()
	defer b.mu.Unlock()
	cutoff := time.Now().Add(-window).UnixMilli()
	firstFresh := 0
	for i, t := range b.callTimestampsMs {
		if t > cutoff {
			firstFresh = i
			break
		}
		firstFresh = i + 1
	}
	if firstFresh > 0 {
		b.callTimestampsMs = b.callTimestampsMs[firstFresh:]
	}
	out := make([]int64, len(b.callTimestampsMs))
	copy(out, b.callTimestampsMs)
	return out
}

// Flush sends the current pending list through the configured FlushFn.
// On error the pending list is preserved so the next tick retries.
//
// If no FlushFn was configured, pending records are silently dropped to
// mirror the TS no-op default.
func (b *BudgetTracker) Flush(ctx context.Context) error {
	b.mu.Lock()
	if len(b.pending) == 0 {
		b.mu.Unlock()
		return nil
	}
	snapshot := make([]PendingSpendRecord, len(b.pending))
	copy(snapshot, b.pending)
	state := BudgetState{
		PolicyID:         b.policyID,
		TotalUsd:         b.totalUsd,
		CallTimestampsMs: append([]int64(nil), b.callTimestampsMs...),
		Pending:          snapshot,
	}
	flushFn := b.flushFn
	b.mu.Unlock()

	if flushFn == nil {
		b.mu.Lock()
		// Drop only what we snapshotted; new records may have landed.
		n := minInt(len(snapshot), len(b.pending))
		b.pending = b.pending[n:]
		b.mu.Unlock()
		return nil
	}
	if err := flushFn(ctx, state); err != nil {
		log.Printf("[BudgetTracker] flush failed for %s: %v", b.policyID, err)
		return err
	}
	b.mu.Lock()
	n := minInt(len(snapshot), len(b.pending))
	b.pending = b.pending[n:]
	b.mu.Unlock()
	return nil
}

// Stop halts the background timer and runs one final flush. Idempotent.
func (b *BudgetTracker) Stop() {
	b.mu.Lock()
	if b.stopped {
		b.mu.Unlock()
		return
	}
	b.stopped = true
	close(b.stopCh)
	b.mu.Unlock()
	b.wg.Wait()
	// Final flush so nothing is lost on shutdown.
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	_ = b.Flush(ctx)
}

func (b *BudgetTracker) loop() {
	defer b.wg.Done()
	ticker := time.NewTicker(b.flushInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			b.mu.Lock()
			has := len(b.pending) > 0
			b.mu.Unlock()
			if !has {
				continue
			}
			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
			_ = b.Flush(ctx)
			cancel()
		case <-b.stopCh:
			return
		}
	}
}

// ServerPushOptions configures the canonical FlushFn that POSTs each
// pending record to POST /v1/agents/:agent_id/spend.
type ServerPushOptions struct {
	CoreURL    string
	AdminKey   string
	AgentID    string
	PolicyID   string
	TenantID   string
	HTTPClient *http.Client
}

// ServerPush builds a FlushFn that POSTs each pending record to the
// server-side spend ledger. See BudgetTrackerOptions.FlushFn.
func ServerPush(opts ServerPushOptions) FlushFn {
	client := opts.HTTPClient
	if client == nil {
		client = &http.Client{Timeout: 10 * time.Second}
	}
	coreURL := trimTrailingSlash(opts.CoreURL)
	target := fmt.Sprintf("%s/v1/agents/%s/spend", coreURL, url.PathEscape(opts.AgentID))
	return func(ctx context.Context, state BudgetState) error {
		for _, rec := range state.Pending {
			body := map[string]interface{}{
				"policy_id":  opts.PolicyID,
				"amount_usd": rec.AmountUsd,
			}
			if rec.ActionID != "" {
				body["action_id"] = rec.ActionID
			}
			buf, err := json.Marshal(body)
			if err != nil {
				return err
			}
			req, err := http.NewRequestWithContext(ctx, http.MethodPost, target, bytes.NewReader(buf))
			if err != nil {
				return err
			}
			req.Header.Set("Content-Type", "application/json")
			if opts.AdminKey != "" {
				req.Header.Set("Authorization", "Bearer "+opts.AdminKey)
			}
			if opts.TenantID != "" {
				req.Header.Set("X-Sauron-Tenant-Id", opts.TenantID)
			}
			resp, err := client.Do(req)
			if err != nil {
				return err
			}
			body2, _ := io.ReadAll(resp.Body)
			resp.Body.Close()
			if resp.StatusCode < 200 || resp.StatusCode >= 300 {
				return fmt.Errorf("POST %s -> %d: %s", target, resp.StatusCode, string(body2))
			}
		}
		return nil
	}
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: sdk/typescript/src/budget-tracker.ts
/**
 * Budget tracker — in-memory ledger of spend and call timestamps. Used
 * by the local invariant evaluator to feed the budget + rate-limit
 * checks without a per-call HTTP roundtrip.
 *
 * Sprint 3 follow-up wires the optional server-side spend ledger:
 * pending records accumulate in `_pending` and a background timer
 * forwards them to `flushFn` every `flushIntervalMs`. The default
 * `BudgetTracker.serverPush()` builder POSTs each pending record to
 * `POST /v1/agents/:agent_id/spend` so the server can hold the
 * authoritative total (closes redteam A3 — local counter tampering).
 *
 * @module
 */

/** One record queued for the next `flush()`. */
export interface PendingSpendRecord {
    /** USD amount as supplied to `record()`. */
    amount_usd: number;
    /** Optional caller-supplied action id (free-form). */
    action_id?: string;
    /** Epoch milliseconds when `record()` was called. */
    timestamp: number;
}

/** Snapshot passed to {@link BudgetTrackerOptions.flushFn}. */
export interface BudgetState {
    /** Policy id this tracker belongs to. */
    policyId: string;
    /** Running USD total. */
    totalUsd: number;
    /** Epoch-millisecond timestamps of recent calls. */
    callTimestampsMs: number[];
    /** Pending records still to be persisted. The default flushFn drains them. */
    pending: PendingSpendRecord[];
}

/** Options for the {@link BudgetTracker} constructor. */
export interface BudgetTrackerOptions {
    /** Policy id whose spend this tracker covers. */
    policyId: string;
    /**
     * Auto-flush interval in ms. Default `30_000` (30s). Pass `0` to
     * disable the timer entirely — callers can still invoke `flush()`
     * manually.
     */
    flushIntervalMs?: number;
    /**
     * Hook invoked from the background timer (and on `stop()`). When
     * absent, pending records are silently dropped after each tick —
     * use `BudgetTracker.serverPush()` to wire the authoritative ledger.
     */
    flushFn?: (state: BudgetState) => Promise<void>;
}

/** Options for {@link BudgetTracker.serverPush}. */
export interface ServerPushOptions {
    /** Base URL of the SauronID core server (no trailing slash). */
    coreUrl: string;
    /** Admin bearer token; sent as `Authorization: Bearer <key>`. */
    adminKey?: string;
    /** Agent id whose ledger row should be incremented. */
    agentId: string;
    /** Policy id this tracker covers. */
    policyId: string;
    /** Override `fetch` (tests / Node < 18). */
    httpFetch?: typeof fetch;
}

/** Lightweight in-memory spend + rate ledger. */
export class BudgetTracker {
    private readonly policyId: string;
    private readonly flushFn: (state: BudgetState) => Promise<void>;
    private readonly flushIntervalMs: number;
    private flushTimer: ReturnType<typeof setInterval> | null = null;

    private totalUsd = 0;
    private readonly callTimestampsMs: number[] = [];
    /** Records added since the last successful `flush()`. */
    private _pending: PendingSpendRecord[] = [];
    private inflightFlush: Promise<void> | null = null;

    constructor(opts: BudgetTrackerOptions) {
        this.policyId = opts.policyId;
        this.flushFn = opts.flushFn ?? (async () => { /* no server wiring */ });
        this.flushIntervalMs = opts.flushIntervalMs ?? 30_000;
        if (this.flushIntervalMs > 0) {
            this.flushTimer = setInterval(() => {
                if (this._pending.length === 0) return;
                void this.flush();
            }, this.flushIntervalMs);
            const t = this.flushTimer as { unref?: () => void };
            if (typeof t.unref === "function") t.unref();
        }
    }

    /**
     * Record one tool invocation. Increments the running total by
     * `amountUsd` (0 if absent), appends a timestamp for rate checks,
     * and queues a `PendingSpendRecord` for the next `flush()`.
     */
    record(amountUsd: number, actionId?: string): void {
        this.totalUsd += amountUsd;
        const now = Date.now();
        this.callTimestampsMs.push(now);
        // Cap history so the array doesn't grow unboundedly. Keeping the
        // last 1024 entries is plenty for a 60s rate window.
        if (this.callTimestampsMs.length > 1024) {
            this.callTimestampsMs.splice(0, this.callTimestampsMs.length - 1024);
        }
        this._pending.push({ amount_usd: amountUsd, action_id: actionId, timestamp: now });
    }

    /** Current spend total in USD. */
    total(): number {
        return this.totalUsd;
    }

    /** Number of records waiting for the next flush. */
    pendingCount(): number {
        return this._pending.length;
    }

    /**
     * Return call timestamps (epoch ms) within the last `windowMs`.
     * Older entries are pruned as a side effect.
     */
    recentCalls(windowMs: number): number[] {
        const cutoff = Date.now() - windowMs;
        // Find the first timestamp inside the window via binary search-ish
        // linear scan from the end (history is append-only ordered).
        let firstFresh = 0;
        for (let i = 0; i < this.callTimestampsMs.length; i++) {
            if (this.callTimestampsMs[i] > cutoff) {
                firstFresh = i;
                break;
            }
            firstFresh = i + 1;
        }
        if (firstFresh > 0) this.callTimestampsMs.splice(0, firstFresh);
        return [...this.callTimestampsMs];
    }

    /**
     * Send the current state through `flushFn`. Drains `_pending` on
     * success; on failure the pending list is preserved so the next
     * tick retries.
     */
    async flush(): Promise<void> {
        // Serialise concurrent flushes — the timer callback can race with a
        // manual `flush()` from `stop()` or user code.
        if (this.inflightFlush) {
            await this.inflightFlush;
            return;
        }
        const snapshot = this._pending.slice();
        if (snapshot.length === 0) return;
        const p = (async () => {
            try {
                await this.flushFn({
                    policyId: this.policyId,
                    totalUsd: this.totalUsd,
                    callTimestampsMs: [...this.callTimestampsMs],
                    pending: snapshot,
                });
                // Drop only the records we sent; new ones may have been
                // appended during the await.
                this._pending.splice(0, snapshot.length);
            } catch (err) {
                // Keep `_pending` so the next tick retries. Log to stderr
                // since this is a background channel.
                // eslint-disable-next-line no-console
                console.warn(
                    `[BudgetTracker] flush failed for ${this.policyId}: ${(err as Error).message}`
                );
            }
        })();
        this.inflightFlush = p;
        try {
            await p;
        } finally {
            this.inflightFlush = null;
        }
    }

    /**
     * Build a `flushFn` that POSTs each pending record to
     * `POST /v1/agents/:agent_id/spend`. Use as:
     *
     * ```ts
     * const flushFn = BudgetTracker.serverPush({
     *   coreUrl, adminKey, agentId, policyId,
     * });
     * const bt = new BudgetTracker({ policyId, flushFn });
     * ```
     */
    static serverPush(opts: ServerPushOptions): (state: BudgetState) => Promise<void> {
        const coreUrl = opts.coreUrl.replace(/\/+$/, "");
        const f = opts.httpFetch ?? (typeof fetch === "function" ? fetch : undefined);
        if (!f) {
            throw new Error("BudgetTracker.serverPush: no fetch available — pass httpFetch");
        }
        const headers: Record<string, string> = { "content-type": "application/json" };
        if (opts.adminKey) headers.authorization = `Bearer ${opts.adminKey}`;
        const url = `${coreUrl}/v1/agents/${encodeURIComponent(opts.agentId)}/spend`;
        return async (state) => {
            for (const rec of state.pending) {
                const body = JSON.stringify({
                    policy_id: opts.policyId,
                    action_id: rec.action_id,
                    amount_usd: rec.amount_usd,
                });
                const r = await f(url, { method: "POST", headers, body });
                if (!r.ok) {
                    throw new Error(
                        `POST ${url} -> ${r.status}: ${await r.text().catch(() => "")}`
                    );
                }
            }
        };
    }

    /**
     * Stop the auto-flush timer and trigger one final flush so no
     * pending records are lost. Idempotent.
     */
    async stop(): Promise<void> {
        if (this.flushTimer) {
            clearInterval(this.flushTimer);
            this.flushTimer = null;
        }
        if (this._pending.length > 0) {
            await this.flush();
        }
    }
}
